File: LeetCode/494. Target Sum/index494.js
/**
 * @param {number[]} nums
 * @param {number} S
 * @return {number}
 */
var findTargetSumWays = function (nums, S) {
  let sum = [];
  nums.forEach((el, index) => {
    if (index) {
      sum[index] = sum[index - 1] + el;
    } else {
      sum[index] = el;
    }
  });
  dp = [];
  dp[0] = {};
  dp[0][nums[0]] = 1;
  dp[0][-nums[0]] = dp[0][-nums[0]] ? 2 : 1;
  for (let i = 1; i < nums.length; i++) {
    dp[i] = {};
    for (let j = -sum[i]; j <= sum[i]; j++) {
      dp[i][j] = (dp[i - 1][j - nums[i]] || 0) + (dp[i - 1][j + nums[i]] || 0);
    }
  }
  return dp[nums.length - 1][S] || 0;
};

findTargetSumWays([1, 2, 1], 0);
findTargetSumWays([0, 0, 0, 0, 0, 0, 0, 0, 1], 1);
findTargetSumWays([1, 1, 1, 1, 1], 3);

File: LeetCode/494. Target Sum/index494.go
package main

import "fmt"

func main() {
	fmt.Println(findTargetSumWays([]int{1, 1, 1, 1, 1}, 3))
	fmt.Println(findTargetSumWays([]int{1}, 1))
}

func findTargetSumWays(nums []int, target int) int {
	dp := make([]map[int]int, len(nums)+1)
	dp[0] = map[int]int{0: 1}

	for i := 0; i < len(nums); i++ {
		dp[i+1] = map[int]int{}
		for j := -1000; j <= 1000; j++ {
			dp[i+1][j] = dp[i][j-nums[i]] + dp[i][j+nums[i]]
		}
	}
	return dp[len(nums)][target]
}
